[p2p] add config to allow setting TCP_NODELAY on TcpStreams

0xK4d1r opened the pull request:
TCP_NODELAY is not on by default in contrary to Golang, so wanted to open a pr for it :)

patrick-ogrady commented:
Congrats on opening the first PR ❤️

patrick-ogrady commented:
Would you mind making this a config (like `read_timeout`, `write_timeout`, etc)? I think it is the right approach to not drop the connection if it can't be disabled for some reason but think we should at least log (probably DEBUG)?

I could see a world where some applications could get a lot better performance by using the Nagle algorithm.

patrick-ogrady commented:
If you get bored, I think the same could eventually be done for linger: https://doc.rust-lang.org/std/net/struct.TcpStream.html#method.set_linger

0xK4d1r commented:
> Would you mind making this a config (like `read_timeout`, `write_timeout`, etc)? I think it is the right approach to not drop the connection if it can't be disabled for some reason but think we should at least log (probably DEBUG)?
> 
> I could see a world where some applications could get a lot better performance by using the Nagle algorithm.

Yeah makes sense, just updated the PR now 👍.

> If you get bored, I think the same could eventually be done for linger: https://doc.rust-lang.org/std/net/struct.TcpStream.html#method.set_linger

Sure I can look into this one too in a follow up PR. I believe its value should be set to `write_timeout` config, right?

patrick-ogrady commented on p2p/src/config.rs:
```diff
@@ -63,6 +63,9 @@ pub struct Config<C: Crypto> {
     /// Duration after which to close the connection if a message cannot be written.
     pub write_timeout: Duration,
 
+    /// Whether or not to disable Nagle's algorithm
```
nit: can you add a few more words here around "that means sending data as soon as possible". end line in a period like others.

patrick-ogrady replied:
Would also be open to also adding (if they have some rationale somewhere):

```
///
/// By default, this is disabled in Golang because.....
```

0xK4d1r replied:
Just added some explanation about it. And on why it is disabled in Golang, I only have is the comment below from the person who wrote the dialer code on HN. I think the explanation I just added is sufficient.

`When I wrote the Go implementation of net.Dial, which I expected to be used for RPC-based systems, it seemed like a no-brainer to set TCP_NODELAY by default.`

patrick-ogrady commented on p2p/src/config.rs:
```diff
@@ -118,6 +121,7 @@ impl<C: Crypto> Config<C> {
             handshake_timeout: Duration::from_secs(5),
             read_timeout: Duration::from_secs(60),
             write_timeout: Duration::from_secs(30),
+            tcp_nodelay: true,
```
nit: please make this default to false

patrick-ogrady commented:
Very close!

> Sure I can look into this one too in a follow up PR. I believe its value should be set to write_timeout config, right?

I would create a separate config but set its value to be the same (in the default) 👍 .

patrick-ogrady commented:
I was reading about `TCP_NODELAY` a tad bit more before merging this...I wonder if it is a better idea to implement this as `Option<>`?

Like with `TCP_LINGER`, not setting `TCP_NODELAY` at all is different than setting it explicitly (will otherwise use OS default behavior).

I think this is less of a concern but I also know that some compile targets don't support this (and might panic or behave unexpectedly). For example, WASI doesn't (not necessarily a goal to work there but also not not a goal lol): https://github.com/rust-lang/rust/blob/19ebdcedf1f5619f8f06c546c7541bb9b54aecda/library/std/src/sys/pal/wasi/net.rs#L150 

My view, when in doubt, is make things as configurable as possible with sane defaults. That would mean making this an `Option` that defaults to `None` and letting devs choose how to use it.

Curious to hear your thoughts!

0xK4d1r commented:
Oh I did not know that it was not supported by some targets. My target is always unix and never thought about this 😄. In that case, making it `Option` makes the most sense indeed. I just converted it to `Option` and updated the comment. Im also thinking now if this should be a target specific code like it only targets `unix` and `windows`?

patrick-ogrady commented:
> Oh I did not know that it was not supported by some targets. My target is always unix and never thought about this 😄. In that case, making it Option makes the most sense indeed.

👍  (I think in all common "server" deployments, it is supported...so probably would've been fine)

> I just converted it to Option and updated the comment. Im also thinking now if this should be a target specific code like it only targets unix and windows?

I think for now, leaving it as is will be fine. I'd prefer to just let anyone using the crate make that decision for themselves.

0xK4d1r commented:
Cool. Btw best of luck on your new journey. Im hoping to contribute more by time 🙂
